src/core/lib/event_engine/grpc_ares_wrapper.h

```diff
@@ -0,0 +1,340 @@
+
+namespace grpc_event_engine {
+namespace experimental {
+
+extern grpc_core::TraceFlag grpc_trace_ares_wrapper;
+
+#define GRPC_ARES_WRAPPER_TRACE_LOG(format, ...)                              \
+  do {                                                                        \
+    if (GRPC_TRACE_FLAG_ENABLED(grpc_trace_ares_wrapper)) {                   \
+      gpr_log(GPR_INFO, "(EventEngine c-ares wrapper) " format, __VA_ARGS__); \
+    }                                                                         \
+  } while (0)
+
+// Base class for the c-ares based DNS lookup functionality. This class is
+// inherited by the concrete leaf classes below which each represents a
+// different type of lookup (i.e. A/AAAA, SRV, TXT). One request object
+// represents one lookup request for a specific name. GrpcAresRequest
+// encapsulates the common logics to initialize and destroy the c-ares channel
+// (one channel for one request). Sets the name servers configuration for the
+// channel. And interacts with c-ares sockets/fds and the EventEngine poller. It
+// also encapsulates logics to start and cancel timers.
+class GrpcAresRequest
+    : public grpc_core::InternallyRefCounted<GrpcAresRequest> {
+ public:
+  ~GrpcAresRequest() override;
+
+  // Cancels an inflight request. Returns true if cancel succeeds and will start
+  // the shutdown process. Returns false if the request is already finishing,
+  // cancelled or shutdown.
+  bool Cancel() ABSL_LOCKS_EXCLUDED(mu_);
+  void Orphan() override {}
+
+ protected:
+  GrpcAresRequest(absl::string_view name, EventEngine::Duration timeout,
+                  std::unique_ptr<GrpcPolledFdFactory> polled_fd_factory,
+                  EventEngine* event_engine);
+
+  // The workhorse function. Gets the live sockets/fds used by c-ares, creates
+  // FdNode if it is not currently tracked in fd_node_list_. Registers the
+  // socket with the poller for read and/or write based on c-ares's demand. And
+  // shutdown and destroys the poller handles whose sockets are no longer in use
+  // by c-ares.
+  // This function is called in every opportunities when there might be a change
+  // to c-ares' sockets for the channel.
+  void Work() ABSL_EXCLUSIVE_LOCKS_REQUIRED(mu_);
+  // Common logic to start the query timeout timer and the ares backup poll
+  // timer. This is only called in each leaf class' Start method.
+  void StartTimers() ABSL_EXCLUSIVE_LOCKS_REQUIRED(mu_);
+  // Common logic to cancel the query timeout timer and the ares backup poll
+  // timer. This is called when the request is cancelled or shutting down.
+  void CancelTimers() ABSL_EXCLUSIVE_LOCKS_REQUIRED(mu_);
+  // These 2 methods are deliberately thread-unsafe and should only be called in
+  // the factory methods of the leaf classes as part of initialization.
+  absl::StatusOr<std::string> ParseNameToResolve();
+  absl::Status InitializeAresOptions(absl::string_view dns_server);
+
+  grpc_core::Mutex mu_;
```

Comment:
Done. Thanks! Go with the getter/setter approach. Reviewing those members is a good practice, some of them only need read-only access from the child class.

Reply:
Sure, "has-a" could work, but there are a few issues as well. Currently the primary states sharing between the base class (`GrpcAresRequest`) and the child classes (`GrpcAres<Type>Request`s) are:
1. Refcounting
2. Mutex
3. OnResolve callback through override a pure virtual member function.
4. ares_channel
5. A few signals, e.g. `cancelled_`, `shutting_down_`.

To modify those into a "has-a" model, i.e. `GrpcAres<Type>Request` has a `GrpcAresRequestCommon` object which encapsulates the shared functionalities, these changes may be necessary:
1. Refcounting: this might be an awkward one. One of the shared functionalities is to pass an OnReadable or OnWritable closure to a poller handle, the closure needs to take a refcount of the `GrpcAres<Type>Request`. So if we don't share refcounting through inheritance, we could maybe pass a raw pointer of `GrpcAres<Type>Request` to `GrpcAresRequestCommon` and it would then invoke `request_->Ref()` and embed the `RefCountedPtr` in the closure. However, this might mean that we need a common interface for all `GrpcAres<Type>Request`s, since `GrpcAresRequestCommon` does not work with any specific `GrpcAres<Type>Request` type. This might also mean that we need to use multiple inheritance for `GrpcAres<Type>Request`s, one for `InternallyRefCounted<>` and one for that interface. Multiple inheritance is also discouraged by the style guide.
2. Mutex: because of the asynchronous nature of many of the shared functionalities, i.e. poller callback and timer callback, the mutex has to be shared in some way between `GrpcAresRequestCommon` and  `GrpcAres<Type>Request`. I could think of to pass a pointer to the `mutex_` member of `GrpcAres<Type>Request` to `GrpcAresRequestCommon`. Note that the Thread Safety Analysis might not behave as expected since it does not support alias analysis: https://clang.llvm.org/docs/ThreadSafetyAnalysis.html#no-alias-analysis.
3. OnResolve callback: this could work through passing a raw pointer of `GrpcAres<Type>Request` to `GrpcAresRequestCommon`. Again it would need a common interface for all `GrpcAres<Type>Request`s.
4. ares_channel: I believe this can live in `GrpcAresRequestCommon` with a getter method.
5. Signals: `cancelled_` or `shutting_down_`: `cancelled_` could live in `GrpcAresRequestCommon` with a getter method. `shutting_down_` could also live in `GrpcAresRequestCommon` but `GrpcAres<Type>Request`s might also need to set it in its static member function. So `GrpcAresRequestCommon` might have a public setter and `GrpcAres<Type>Request`s might have a private setter which will invoke the public setter.

For the point 1 about Refcounting, there might be another approach to only make `GrpcAresRequestCommon` InternallyRefCounted, but that may come with its own complexity.

Honestly, if we really don't like the implementation inheritance, we could also try to extract some of the functionalities in `GrpcAresRequest` into free functions, e.g. initialize an ares_channel, set dns server, process fd_node_list. There will be some code duplications, e.g. timers, OnReadable OnWritable callbacks, but won't be a lot. And we can avoid the complexities with either implementation inheritance or composition. There might be a small interface inheritance for `Cancel` and `Orphan` methods exposing to the EE DNSResolver.

Reply:
Let's take a step back.  I think the overall problem here is that we were supposed to be writing new code, and instead it looks like we've basically copied the existing horrible code and just made a few small tweaks to it.  The existing code has evolved over a long period of time, through many authors, and has never had a clear architecture imposed on it.  It is a rat's nest of implicit dependencies, complex and undocumented interactions between base class and subclasses, and tons of shared state.  No one can possibly understand this, and I think we need to do better.  The more we talk about this, the less happy I am with it.

What I want to see here is an architecture where there are clearly defined components, each with obvious and well documented responsibility and scope, and each providing a no-larger-than-absolutely-necessary API surface that makes it obvious how to use it.  The goal here is to make it easy for readers to understand which objects are responsible for what and how they interact with each other.  We need this code to be maintainable enough that members of the team can work with it without having to first spend days trying to understand the code.

I do not want to see any code where we are sharing a mutex between a base class and subclasses.  That sort of pattern is a horrible code smell; it can sometimes be necessary in performance-sensitive code, but this code does not fall into that category, so there's no justification for it here.

I do not want to see so much shared state between the base class and subclasses.  The style gude specifically recommends [against overusing implementation inheritence](https://google.github.io/styleguide/cppguide.html#Inheritance), and I think this code is a great example of why it says that.

I think we should step back and try to design a cleaner way of structuring this.  Here's one possible approach:

1. A polling object, which deals with all of the interaction with the EventEngine polling code, the fds, etc.
   - The API exposed by this object should allow the caller to start a particular c-ares query (e.g., an A/AAAA query, a TXT query, or an SRV query) and specify a callback to be invoked with the result of that query.  Note that this callback should be in a form that does not care what type of object it is calling back into.
   - The OnReadable and OnWritable callbacks should be purely internal to this object.  Every time one of those callbacks is invoked, we should do the appropriate reading and writing.  When we determine that a request is complete, we invoke the on-complete callback for that request.
   - As per the other comment thread, this object might be the top-level `DNSResolver` object itself.  Or, if we elect to keep this state separate for each individual query, then it should be a separate object that we create for each query.
2. The query type objects, each of which represents a query of a particular type (A/AAAA, TXT, or SRV).
   - If the polling object is actually the `DNSResolver` object, then each of these higher-level objects will hold a ref to the `DNSResolver`; otherwise, each of these higher-level objects should create its own instance of the polling object.
   - The query-type objects should tell the polling object what type of c-ares query to start and provide an on-complete callback to be invoked by the polling object when the query is complete.
   - The query-type objects will be responsible for tracking the task handles and dealing with cancellation.  If we continue to do cancellation on a per-query basis, then when a query-type object is cancelled, it can propagate that cancellation to its polling object.
   - The on-complete callback will be responsible for doing query-type-specific response parsing and invoking the on-complete callback provided by the caller of the EE API.

I think in this approach, there is no need for any inheritence or any mutex sharing, and the interface between the polling code and the query-type objects is a very small, well-defined, easy-to-understand interface.

Why can't something like this work?